08-optional-types.Rmd

```diff
@@ -0,0 +1,344 @@
+```{r, include = FALSE}
+ottrpal::set_knitr_image_path()
+```
+
+# Optional types
+
+WDL supports defining optional variables, which are denoted with a question mark next to the type declaration. These variables may or may not defined. Optional types are powerful, but they can quickly cause problems in complicated workflows. This is especially the case when working with scattered tasks. In this section, we'll go over how optional variables can be used, and where they can cause problems.
+
+## Optional inputs
+The most common use case for optional variables are optional inputs for the user to provide some sort of file, for example, a reference genome or a metadata file. Sometimes, you will want optional inputs to fall back on a known file or value.
+
+A good place to use optional types in our workflow would be in the `bwa mem` task. `bwa mem` on default settings may attempt to use too many threads, which could cause this task to fail if it ran on weaker hardware such as a laptop. In order to make `bwa mem` run as expected on different backends, we can set `Int? bwa_mem_threads` as a workflow-level optional input. Making it optional means that users who do not want to think about threads -- perhaps they're using an HPC or other powerful backend -- do not have to worry about filling in that value. In the workflow body, when we call `bwa mem`, the value of `bwa_mem_threads` is passed to the `BwaMem` task.
+
+```
+  call BwaMem {
```

Comment:
Let's incorporate a snippet of the workflow to show where `Int? bwa_mem_threads` is being defined. I think we should show code snippets where things are being changed. I was a little bit confused where the optional type was in this code block.

Reply:
Thank you for bringing this up! This part especially needed some extra attention.